core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/UnionSailSource.java

```diff
@@ -64,8 +64,8 @@ public void prepare() throws SailException {
 
 	@Override
 	public void flush() throws SailException {
-		primary.flush();
 		additional.flush();
+		primary.flush();
```

Comment:
@jeenbroekstra 

Here is the essence of the problem. Flushing either of `additional` or `primary` will flush the changes to the underlying memory store and also move to the next snapshot. The fact that it moves to the next snapshot will make the changes visible to other transactions, but this is problematic since `additional` and `primary` both operate on the same snapshot version.

The current algorithm can be flattened as something close to:

1. flush changes from `primary`
1. make current snapshot visible to other transactions
1. flush changes from `additional` to the visible snapshot
1. redundantly make the current snapshot visible again


It should be something along the lines of:

1. flush changes from `primary`
1. flush changes from `additional`
1. make current snapshot visible to other transactions

Swapping `primary` and `additional` like I did here doesn't fix the underlying issue, just masquerades it.